checkpolicy,libsepol: add prefix/suffix matching to filename type transitions

JurajMarcin opened the pull request:
Currently, filename transitions are stored separately from other type
enforcement rules and only support exact name matching. However, in
practice, the names contain variable parts. This leads to many
duplicated rules in the policy that differ only in the part of the name,
or it is even impossible to cover all possible combinations.

This series implements equivalent changes made by this kernel patch
series [1].

First, this series of patches moves the filename transitions to be part
of the avtab and avrule structures. This not only makes the
implementation of prefix/suffix matching and future enhancements easier,
but also reduces the technical debt regarding the filename transitions.
Next, the last three patches implement the support for prefix/suffix
name matching itself by extending the structures added in previous
patches in this series and adding the support to CIL in the last of the
triple.

Even though, moving everything to avtab increases the memory usage and
the size of the binary policy itself and thus the loading time, the
ability to match the prefix or suffix of the name will reduce the
overall number of rules in the policy which should mitigate this issue.

[1]: https://lore.kernel.org/selinux/20230601170302.nrhuay2wh44g6sh4@jmarcin-t14s-01/


Changelog:

v2:
- rebased to the latest upstream main
- added syntax examples to commit messages (suggested by Jim)
- removed the "match_" prefix from keywords and dropped the
  "match_exact" keyword (suggested by Jim, discussed with Ondrej)
- fixed the bug Jim found [2] where the output to conf and CIL did not
  work
- reworked avtab entry skipping when writing older versions of the
  policy
- fixed bug where expand_terule_helper() skipped next entries after
  filename transition entry
- fixed memory leak in filenametr_destroy() found by Christian in the
  kernel version of the code [3]

[2]: https://lore.kernel.org/selinux/CAP+JOzS1nhxF8EttikRtYBW2QtW=ck+P4mC2dovgP4kp-g9WQQ@mail.gmail.com/
[3]: https://lore.kernel.org/selinux/CAJ2a_DecaSa1Z7HiRs4sGUGx3AyKwhFA41EhAo+BViv-qA9ndg@mail.gmail.com/

JurajMarcin committed:
checkpolicy, libsepol: move transition to separate structure in avtab

To move filename transitions to be part of avtab, we need to create
space for it in the avtab_datum structure which holds the rule for
a certain combination of stype, ttype and tclass.

As only type transitions have a special variant that uses a filename, it
would be suboptimal to add a (mostly empty) pointer to some structure to
all avtab rules.

Therefore, this patch adds a new structure to the avtab_datum and moves
the otype of the transition to this structure. In the next patch, this
structure will also hold filename transitions for the combination of
stype, ttype and tclass.

Reviewed-by: Ondrej Mosnacek <omosnace@redhat.com>
Signed-off-by: Juraj Marcin <juraj@jurajmarcin.com>

JurajMarcin committed:
checkpolicy, libsepol: move filename transitions to avtab

Currently, filename transitions are stored separately from other type
enforcement rules. This leads to possibly sub-optimal performance and
makes further improvements cumbersome.

This patch adds a symbol table with filename transitions to the
transition structure added to avtab in the previous patch. It also
implements functions required for reading and writing filename
transitions (either binary or source formats) and updates the code for
expanding attributes. Last but not least, it updates the conflict check
in the conditional avtab to account for empty transitions in the
non-conditional avtab.

These changes are expected to cause higher memory usage, as now there
needs to be a filename transition structure for every stype. This patch
effectively undoes most of the commit 42ae834a ("libsepol,checkpolicy:
optimize storage of filename transitions"), but this will be mitigated
by providing support for matching prefix/suffix of the filename for
filename transitions in future patches which will reduce to need to have
so many of them.

Reviewed-by: Ondrej Mosnacek <omosnace@redhat.com>
Signed-off-by: Juraj Marcin <juraj@jurajmarcin.com>

JurajMarcin committed:
checkpolicy, libsepol: move filename transition rules to avrule

Similarly to the previous patch, filename transition rules are stored
and parsed separately from other type enforcement rules. Moving them to
avrule makes it consistent with the filename transitions in avtab and
makes future improvements easier to implement.

This patch adds an optional object name attribute to the avrule
structure and uses this new attribute to move filename transition rules
to avrule. It also updates functions for parsing type enforcement rules
to accept rules with a filename as their last argument (filename
transition rules), separate functions for parsing filename transitions
are therefore no longer needed.

Reviewed-by: Ondrej Mosnacek <omosnace@redhat.com>
Signed-off-by: Juraj Marcin <juraj@jurajmarcin.com>

JurajMarcin committed:
libsepol: implement new kernel binary format for avtab

Implement a new binary policy format that closely matches the new
internal representation introduced in the previous patch.

This patch bumps the maximum kernel policy version and implements
reading/writing functions such that kernel binary policy structure
matches internal representation.

These changes can cause the binary policy to grow in size due to
effectively undoing the benefits of the commit 8206b8cb ("libsepol:
implement POLICYDB_VERSION_COMP_FTRANS "), but this will be mitigated by
adding the prefix/suffix support as described in the previous patch.

Reviewed-by: Ondrej Mosnacek <omosnace@redhat.com>
Signed-off-by: Juraj Marcin <juraj@jurajmarcin.com>

JurajMarcin committed:
libsepol: implement new module binary format of avrule

Implement a new module policy format that closely matches the new
internal representation of avrule introduced in the previous patch.

This patch bumps the maximum module policy version and implements
reading/writing functions such that the module binary policy structure
matches its internal representation, namely, the object name attribute
used for the filename transition rules.

These changes have no significant effect on the size of the module
policy file (tested with Fedora policy).

Reviewed-by: Ondrej Mosnacek <omosnace@redhat.com>
Signed-off-by: Juraj Marcin <juraj@jurajmarcin.com>

WOnder93 commented on libsepol/src/avtab.c:
```diff
@@ -662,3 +674,179 @@ int avtab_read(avtab_t * a, struct policy_file *fp, uint32_t vers)
+		}
+		new_fnt->next = datum->trans->fnts;
+		datum->trans->fnts = new_fnt;
+		return 0;
```
`return SEPOL_OK`

Also, I think it would be a bit more readable if you "deduplicate" the return by writing this part as:
```
if (datum) {
    ...
} else {
    ...
}
return SEPOL_OK;
```

WOnder93 commented on libsepol/src/avtab.c:
```diff
@@ -662,3 +674,179 @@ int avtab_read(avtab_t * a, struct policy_file *fp, uint32_t vers)
+				return -1;
+		}
+	}
+
```
Maybe drop this empty line?

WOnder93 commented on libsepol/src/expand.c:
```diff
@@ -1753,6 +1756,11 @@ static int expand_terule_helper(sepol_handle_t * handle,
+			 * TODO: if old type is zero, then it should insert new
+			 * rule
+			 */
+
```
If I understand it correctly, this patch breaks this case and the next one fixes it? If so (and it can't be worked around here in a simple way), maybe it would be better to merge those two patches into one.

JurajMarcin replied:
Fixed in the last update, added code to check if there actually is any (normal) type transition.

WOnder93 commented on libsepol/src/kernel_to_cil.c:
```diff
@@ -1803,7 +1822,7 @@ static int map_avtab_write_helper(avtab_key_t *key, avtab_datum_t *datum, void *
+		rule = avtab_node_to_str(pdb, key, datum, strs);
 		if (!rule) {
 			rc = -1;
 			goto exit;
```
Since we may now have more than one "str" per rule, I would suggest to move the `strs` adding fully into the helper function and rename it (e.g. to `avtab_node_to_strs`) + change the return type to int.

Same for `kernel_to_conf.c`.

WOnder93 commented on libsepol/include/sepol/policydb/policydb.h:
```diff
@@ -161,19 +161,6 @@ typedef struct role_allow {
 	struct role_allow *next;
 } role_allow_t;
 
-/* filename_trans rules */
-typedef struct filename_trans_key {
-	uint32_t ttype;
-	uint32_t tclass;
-	char *name;
-} filename_trans_key_t;
-
-typedef struct filename_trans_datum {
-	ebitmap_t stypes;
-	uint32_t otype;
-	struct filename_trans_datum *next;
-} filename_trans_datum_t;
-
```
I believe you can also remove the `filename_trans` and `filename_trans_count` members of `struct policydb` in this patch.

WOnder93 commented on libsepol/src/write.c:
```diff
@@ -314,7 +318,19 @@ static int avtab_write(struct policydb *p, avtab_t * a, struct policy_file *fp)
 		nel = a->nel;
 	} else {
 		/* New avtab format.  nel is good to go. */
```
This comment is not really true after this patch so should be removed.

WOnder93 commented on libsepol/src/write.c:
```diff
@@ -314,7 +318,19 @@ static int avtab_write(struct policydb *p, avtab_t * a, struct policy_file *fp)
-		nel = cpu_to_le32(a->nel);
+		nel = a->nel;
+		/*
+		 * entires containing only filename transitions are skipped and
```
entires -> entries

WOnder93 commented on libsepol/src/write.c:
```diff
@@ -314,7 +318,19 @@ static int avtab_write(struct policydb *p, avtab_t * a, struct policy_file *fp)
 		nel = a->nel;
```
Is this `oldvers` branch possibly affected by the avtab changes? Perhaps we need to adjust the `nel` here as well? (Alternatively perhaps change `expand_avtab()` to remove the named transitions - but only if this is its only user.)

JurajMarcin replied:
Fixed by decrementing `nel` in `avtab_write_item()` during the first pass.

WOnder93 commented on libsepol/src/write.c:
```diff
@@ -358,6 +374,299 @@ static int avtab_write(struct policydb *p, avtab_t * a, struct policy_file *fp)
+				}
+			}
+		}
+
```
Unnecessary empty line.

WOnder93 commented on libsepol/src/write.c:
```diff
@@ -2381,11 +2578,11 @@ int policydb_write(policydb_t * p, struct policy_file *fp)
-			if (p->filename_trans)
-				WARN(fp->handle, "Discarding filename type transition rules");
+			WARN(fp->handle,
+			     "Discarding filename type transition rules");
```
This should look into the avtab and only print the warning if there actually are any ftrans rules.

WOnder93 commented on libsepol/src/policydb_validate.c:
```diff
@@ -936,6 +936,15 @@ static int validate_avrules(sepol_handle_t *handle, const avrule_t *avrule, int
+		    avrule->specified & AVRULE_TRANSITION &&
+		    avrule->object_name)
+			goto bad;
+
```
This shouldn't be needed - both normal and filename transitions support the RULE_SELF flag, thus the switch statement below already validates it sufficiently.

WOnder93 commented on libsepol/src/expand.c:
```diff
@@ -1732,6 +1644,33 @@ static int expand_terule_helper(sepol_handle_t * handle,
+			if (rc == SEPOL_EEXIST) {
+				ERR(handle, "conflicting filename transition %s %s:%s \"%s\": %s vs %s",
+				    p->p_type_val_to_name[avkey.source_type -
+							  1],
```
I'd suggest to keep these on the same line for better readability. We don't enforce the 80-char rule in SELinux userspace and even the kernel's checkpatch.pl now allows up to 100 chars per line.

WOnder93 commented on libsepol/src/write.c:
```diff
@@ -102,6 +102,50 @@ static uint16_t spec_order[] = {
+		if (items != 1)
+			return -1;
+	}
+
```
Unnecessary empty line.

WOnder93 commented on libsepol/src/expand.c:
```diff
@@ -1729,24 +1644,62 @@ static int expand_terule_helper(sepol_handle_t * handle,
 			}
+			/*
+			 * conditional avtab does not contain filename
+			 * transtions, no need to check for empty
```
typo "transtions"

Also I'd suggest to clarify the last part - "no need to check for otype == 0", perhaps?

WOnder93 reviewed (changes requested):
Just two minor suggestions regarding the newly added comment, but otherwise LGTM (including the commit messages + cover text). :+1:

WOnder93 commented on libsepol/src/avtab.c:
```diff
@@ -418,6 +447,88 @@ void avtab_hash_eval(avtab_t * h, char *tag)
+		return rc;
+	trans->otype = le32_to_cpu(buf32[0]);
+
+
```
Double empty line.

WOnder93 commented on checkpolicy/test/dispol.c:
```diff
@@ -130,10 +147,16 @@ static int render_av_rule(avtab_key_t * key, avtab_datum_t * datum, uint32_t wha
+				render_type(datum->trans->otype, p, fp);
+				fprintf(fp, ";\n");
+			}
+			render_name_trans_args_t args = {key, p, fp};
```
Probably better to use explicit field initialization to avoid relying on the ordering (`{ .key = key, .p = p, .fp = fp }`).

WOnder93 commented on libsepol/src/expand.c:
```diff
@@ -1729,24 +1644,62 @@ static int expand_terule_helper(sepol_handle_t * handle,
 			}
+			/*
+			 * conditional avtab does not contain filename
+			 * transitions, no need for otype == 0
```
"no need to check for"?

WOnder93 commented on libsepol/src/kernel_to_cil.c:
```diff
@@ -1757,33 +1766,38 @@ static char *avtab_node_to_str(struct policydb *pdb, avtab_key_t *key, avtab_dat
+			if (rc < 0)
+				return rc;
+		}
+		name_trans_to_strs_args_t args = {pdb, strs, flavor, src, tgt, class};
```
`{ .field = field, ... }`

WOnder93 commented on libsepol/src/kernel_to_conf.c:
```diff
@@ -1729,32 +1739,36 @@ static char *avtab_node_to_str(struct policydb *pdb, avtab_key_t *key, avtab_dat
 		}
-
-		rule = create_str("%s %s %s:%s %s;", 5, flavor, src, tgt, class, perms);
+		name_trans_to_strs_args_t args = {pdb, strs, flavor, src, tgt, class};
```
`{ .field = field, ... }`

WOnder93 commented on libsepol/src/avtab.c:
```diff
@@ -630,3 +755,186 @@ int avtab_read(avtab_t * a, struct policy_file *fp, uint32_t vers)
+	}
+
+	if (!datum->trans->name_trans.table ||
+	    !datum->trans->name_trans.table->size) {
```
This would ideally be a helper function (see https://gitlab.com/JurajMarcin/selinux-kernel/-/merge_requests/1#note_1170768134).

WOnder93 commented on libsepol/src/write.c:
```diff
@@ -102,6 +102,58 @@ static uint16_t spec_order[] = {
 };
 
+
+
```
Extraneous empty lines.

WOnder93 commented on libsepol/src/write.c:
```diff
@@ -302,8 +375,26 @@ static int avtab_write(struct policydb *p, avtab_t * a, struct policy_file *fp)
+			for (i = 0; i < a->nslot; i++) {
+				for (cur = a->htable[i]; cur; cur = cur->next) {
+					if (cur->key.specified
+					    & AVTAB_TRANSITION &&
```
Please put parentheses around the `... & ...`. It's better to be explicit when mixing `&` and `&&` in a single expression.

WOnder93 commented on libsepol/src/avtab.c:
```diff
@@ -304,6 +315,21 @@ avtab_ptr_t avtab_search_node_next(avtab_ptr_t node, int specified)
+static int avtab_trans_destroy_helper(hashtab_key_t k, hashtab_datum_t d,
+				      void *a)
+{
+	(void)(a);
```
In the SELinux userspace codebase we normally mark unused arguments with `__attribute__ ((unused))` (in the function prototype).

WOnder93 commented on libsepol/src/policydb_validate.c:
```diff
@@ -817,15 +817,42 @@ static int validate_xperms(const avtab_extended_perms_t *xperms)
+
+static int validate_name_trans_helper(hashtab_key_t k, hashtab_datum_t d, void *a)
+{
+	(void)(k);
```
Also `__attribute__ ((unused))`.

JurajMarcin committed:
checkpolicy, libsepol: add prefix/suffix support to kernel policy

Currently, filename type transitions support only exact name matching.
However, in practice, the names contain variable parts. This leads to
many duplicated rules in the policy that differ only in the part of the
name, or it is even impossible to cover all possible combinations.

This patch extends the filename type transitions structures to include
new types of filename transitions - prefix and suffix filename
transitions. It also implements the reading and writing of those rules
in the kernel binary policy format together with increasing its version.

Reviewed-by: Ondrej Mosnacek <omosnace@redhat.com>
Signed-off-by: Juraj Marcin <juraj@jurajmarcin.com>

JurajMarcin commented on checkpolicy/policy_define.c:
```diff
@@ -2358,6 +2370,13 @@ static int avrule_cpy(avrule_t *dest, const avrule_t *src)
+		dest->object_name = strdup(src->object_name);
+		if (!dest->object_name) {
+			yyerror("out of memory");
+			return 01;
```
Return `-1` instead of `01`.

JurajMarcin committed:
checkpolicy, libsepol: add prefix/suffix support to module policy

This patch extends the structures for module and base policy (avrule_t)
to support prefix/suffix transitions. In addition to this, it implements
the necessary changes to functions for reading and writing the binary
policy as well as parsing the policy conf.

Reviewed-by: Ondrej Mosnacek <omosnace@redhat.com>
Signed-off-by: Juraj Marcin <juraj@jurajmarcin.com>

JurajMarcin committed:
libsepol/cil: add support for prefix/suffix filename transtions to CIL

This patch implements the support for prefix/suffix filename transitions
in CIL structures as well as to CIL policy parser.

Reviewed-by: Ondrej Mosnacek <omosnace@redhat.com>
Signed-off-by: Juraj Marcin <juraj@jurajmarcin.com>
